[clang] Implement function pointer signing and authenticated function calls

ahmedbougacha opened the pull request:
The functions are currently always signed/authenticated with zero discriminator.

Co-Authored-By: John McCall <rjmccall@apple.com>

ahatanak committed:
[clang] Implement function pointer signing.

Co-Authored-By: John McCall <rjmccall@apple.com>

ahatanak committed:
Remove unused functions and add test case

kovdan01 commented on clang/lib/CodeGen/CGPointerAuthInfo.h:
```diff
@@ -0,0 +1,96 @@
+  PointerAuthenticationMode AuthenticationMode : 2;
+  bool IsIsaPointer : 1;
+  bool AuthenticatesNullValues : 1;
+  unsigned Key : 28;
```
Do we need 28 bits for `Key`? I do not think such manual "padding" is needed, 2 bits should be enough for 4 keys (IA, IB, DA, DB)

ahatanak replied:
Currently we only need 2 bits, but we'll need more when we add support for soft keys. I changed the size to 4 to match the size in `PointerAuthSchema`.

kovdan01 commented on clang/lib/CodeGen/CGPointerAuthInfo.h:
```diff
@@ -0,0 +1,96 @@
+    return AuthenticationMode == PointerAuthenticationMode::SignAndAuth;
+  }
+
+  bool operator!=(const CGPointerAuthInfo &Other) const {
```
nit: `!=` and `==` better look as free functions since their operands order is not significant

kovdan01 commented on clang/lib/CodeGen/CodeGenFunction.cpp:
```diff
@@ -3041,3 +3046,20 @@ llvm::Value *CodeGenFunction::emitBoolVecConversion(llvm::Value *SrcVec,
   return Builder.CreateShuffleVector(SrcVec, ShuffleMask, Name);
 }
+
+void CodeGenFunction::EmitPointerAuthOperandBundle(
```
`PascalCase` for variables in this function - surrounding code already uses this naming convention

kovdan01 commented on clang/lib/CodeGen/CodeGenFunction.cpp:
```diff
@@ -3041,3 +3046,20 @@ llvm::Value *CodeGenFunction::emitBoolVecConversion(llvm::Value *SrcVec,
+  if (!pointerAuth.isSigned())
+    return;
+
+  auto key = Builder.getInt32(pointerAuth.getKey());
```
```suggestion
  auto *Key = Builder.getInt32(pointerAuth.getKey());
```

kovdan01 commented on clang/lib/CodeGen/CGCall.h:
```diff
@@ -104,10 +109,13 @@ class CGCallee {
-  CGCallee(const CGCalleeInfo &abstractInfo, llvm::Value *functionPtr)
+  CGCallee(
+      const CGCalleeInfo &abstractInfo, llvm::Value *functionPtr,
+      const CGPointerAuthInfo &pointerAuthInfo = /*FIXME*/ CGPointerAuthInfo())
```
Could you please provide additional context for this FIXME?

ahatanak replied:
The `CGPointerAuthInfo` parameter is going to be mandatory later.

ahatanak replied:
Yes, that's correct.

kovdan01 commented on clang/lib/CodeGen/CGPointerAuth.cpp:
```diff
@@ -28,6 +28,24 @@
+/// Return the abstract pointer authentication schema for a pointer to the given
+/// function type.
+CGPointerAuthInfo CodeGenModule::getFunctionPointerAuthInfo(QualType T) {
+  auto &Schema = getCodeGenOpts().PointerAuth.FunctionPointers;
```
```suggestion
  const auto &Schema = getCodeGenOpts().PointerAuth.FunctionPointers;
```

kovdan01 commented on clang/lib/CodeGen/CodeGenModule.h:
```diff
@@ -26,6 +26,7 @@
 #include "clang/Basic/LangOptions.h"
 #include "clang/Basic/Module.h"
 #include "clang/Basic/NoSanitizeList.h"
+#include "clang/Basic/PointerAuthOptions.h"
```
This include looks unneeded since further forward declaration `class PointerAuthSchema;` does the job

kovdan01 commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -1458,6 +1458,39 @@ static void setPGOUseInstrumentor(CodeGenOptions &Opts,
+}
+
+static bool parsePointerAuthOptions(PointerAuthOptions &Opts,
+                                    ArgList &Args,
```
```suggestion
                                    const ArgList &Args,
```

ahatanak replied:
I removed `Args` as it wasn't used.

kovdan01 commented on clang/include/clang/Basic/LangOptions.h:
```diff
@@ -346,6 +346,8 @@ class LangOptionsBase {
     BKey
   };
 
+  using PointerAuthenticationMode = ::clang::PointerAuthenticationMode;
```
What does this `using` directive serve for? I've deleted it locally and the build still succeeds.

ahatanak replied:
I deleted it as it's not needed for this PR.

kovdan01 commented on clang/include/clang/Basic/PointerAuthOptions.h:
```diff
@@ -14,10 +14,146 @@
+
+  uint16_t getConstantDiscrimination() const {
+    assert(getOtherDiscrimination() == Discrimination::Constant);
+    return (uint16_t)ConstantDiscriminator;
```
Is the manual cast really needed? The following works for me just fine:
```suggestion
    return ConstantDiscriminator;
```

kovdan01 commented on clang/include/clang/Basic/PointerAuthOptions.h:
```diff
@@ -14,10 +14,146 @@
+    case Kind::None:
+      llvm_unreachable("calling getKey() on disabled schema");
+    case Kind::ARM8_3:
+      return unsigned(getARM8_3Key());
```
Consider using `llvm::to_underlying` from llvm/ADT/STLForwardCompat.h:
```suggestion
      return llvm::to_underlying(getARM8_3Key());
```

kovdan01 commented on clang/lib/CodeGen/CGCall.h:
```diff
@@ -128,12 +136,12 @@ class CGCallee {
   static CGCallee forDirect(llvm::Constant *functionPtr,
                             const CGCalleeInfo &abstractInfo = CGCalleeInfo()) {
-    return CGCallee(abstractInfo, functionPtr);
+    return CGCallee(abstractInfo, functionPtr, CGPointerAuthInfo());
```
Do we need to pass the default value manually? Can't we just leave this "as is"?

ahatanak replied:
Yes. I've fixed that.

kovdan01 commented on clang/lib/CodeGen/CGPointerAuthInfo.h:
```diff
@@ -0,0 +1,96 @@
+
+class CGPointerAuthInfo {
+private:
+  PointerAuthenticationMode AuthenticationMode : 2;
```
`enum class PointerAuthenticationMode` is defined in "clang/Basic/LangOptions.h". Consider including that explicitly not to rely on transitive includes.

ahatanak commented on clang/include/clang/Basic/PointerAuthOptions.h:
```diff
@@ -14,10 +14,146 @@
+
+private:
+  Kind TheKind : 2;
+  unsigned IsAddressDiscriminated : 1;
```
I don't think it makes any difference on macOS or Linux, but mixing bool bitfields and unsigned bitfields can increase the size of the struct on Windows (see https://learn.microsoft.com/en-us/cpp/c-language/storage-and-alignment-of-structures?view=msvc-170). The underlying types of the enums used here are all unsigned.

I guess we should use `unsigned` as the underlying type of bitfields in `class CGPointerAuthInfo` too.

kovdan01 replied:
> I guess we should use `unsigned` as the underlying type of bitfields in `class CGPointerAuthInfo` too.

This works for me, thanks

ahatanak committed:
Address review comments

- Remove unneeded declarations.
- Fix includes.
- Use `llvm::to_underlying`.
- Fix variable names.
- Fix size and type of bit-fields.
- Add test case for `err_drv_ptrauth_not_supported`.

ahatanak commented on clang/lib/CodeGen/CGPointerAuth.cpp:
```diff
@@ -75,3 +93,36 @@ CodeGen::getConstantSignedPointer(CodeGenModule &CGM,
+                                                  llvm::Type *Ty) {
+  const FunctionDecl *FD = cast<FunctionDecl>(GD.getDecl());
+
+  // Annoyingly, K&R functions have prototypes in the clang AST, but
```
I removed this piece of code as we can't test K&R functions until we have function pointer type discrimination.

ahatanak committed:
Declare operator== as a friend function instead of a member function

asl commented:
@ahatanak Looks like there are some conflicts that should be resolved

kovdan01 commented on clang/test/CodeGen/ptrauth-function.c:
```diff
@@ -0,0 +1,28 @@
+void test_call();
+
+// CHECK-LABEL: define void @test_direct_call()
+void test_direct_call() {
```
Are indirect calls tested? I see clang/test/CodeGen/ptrauth-function-lvalue-cast.c testing indirect call via `char *ptr` casted to desired function type, but I don't see a test when we just have a signed function pointer (for example, passed by an argument to a function) and call the function that way. Please let me know if I miss smth.

kovdan01 commented on clang/lib/CodeGen/CodeGenFunction.cpp:
```diff
@@ -3041,3 +3046,20 @@ llvm::Value *CodeGenFunction::emitBoolVecConversion(llvm::Value *SrcVec,
+  if (!PointerAuth.isSigned())
+    return;
+
+  auto Key = Builder.getInt32(PointerAuth.getKey());
```
```suggestion
  auto *Key = Builder.getInt32(PointerAuth.getKey());
```

kovdan01 commented on clang/lib/CodeGen/CodeGenModule.h:
```diff
@@ -938,6 +939,32 @@ class CodeGenModule : public CodeGenTypeCache {
+
+  CGPointerAuthInfo getFunctionPointerAuthInfo(QualType T);
+
+  llvm::Constant *getConstantSignedPointer(llvm::Constant *Pointer,
```
This member function declaration seems to be unused and not corresponding to any definition, so it can probably be removed.

ahatanak committed:
Address review comments

- Remove unused diagnostic.
- Change the size of Key fields to 2 bits.
- Stop passing the default argument to CGCallee's constructor.
- Remove unused functions and variables.
- Add a test for indirect calls.
- Run tests on linux too.

ahatanak commented on clang/include/clang/Basic/DiagnosticDriverKinds.td:
```diff
@@ -351,6 +351,9 @@ def err_drv_omp_host_ir_file_not_found : Error<
 def err_drv_omp_host_target_not_supported : Error<
   "target '%0' is not a supported OpenMP host target">;
+def err_drv_ptrauth_not_supported : Error<
+  "target '%0' does not support native pointer authentication">;
```
I removed the diagnostic. This was needed before `TargetSpecific` was added to the driver options.

ahatanak commented on clang/lib/CodeGen/CGPointerAuth.cpp:
```diff
@@ -61,3 +79,28 @@ CodeGen::getConstantSignedPointer(CodeGenModule &CGM, llvm::Constant *Pointer,
+/// functionType.
+llvm::Constant *CodeGenModule::getFunctionPointer(llvm::Constant *Pointer,
+                                                  QualType FunctionType,
+                                                  GlobalDecl GD) {
```
No, it isn't. I've removed it. It will be needed when we add support for function-type based signing later.

ahatanak commented on clang/lib/CodeGen/CGPointerAuth.cpp:
```diff
@@ -61,3 +79,28 @@ CodeGen::getConstantSignedPointer(CodeGenModule &CGM, llvm::Constant *Pointer,
                                       OtherDiscriminator);
 }
+
+/// If applicable, sign a given constant function pointer with the ABI rules for
```
I removed the comment.

ahatanak commented on clang/test/Driver/ptrauth.c:
```diff
@@ -0,0 +1,9 @@
+// RUN: %clang -target arm64-apple-macosx -fptrauth-calls -c %s -### 2>&1 | FileCheck %s --check-prefix PTRAUTH_CALLS
```
I removed the file.

ahatanak commented on clang/test/CodeGen/ptrauth-function-attributes.c:
```diff
@@ -0,0 +1,13 @@
+// RUN: %clang_cc1 -triple arm64-apple-ios                    -emit-llvm %s  -o - | FileCheck %s --check-prefixes=ALL,OFF
```
I've added RUN lines for `aarch64-linux-gnu`.

ahatanak commented on clang/test/CodeGen/ptrauth-function-init-fail.c:
```diff
@@ -0,0 +1,5 @@
+
+void f(void);
+
+int *pf = (int *)&f + 1; // expected-error{{cannot compile this static initializer yet}}
```
As commented in the test case, adding a non-zero constant to a signed pointer is unrepresentable in relocations. I agree that the diagnostic should provide more information and should inform the users that the initialization failed because of this specific limitation.

I'll see if I can improve the error message later.

kovdan01 replied:
> I'll see if I can improve the error message later.

Thanks. I'm OK with having the diagnostic "as is" for this PR and submitting a patch with message improvement as another PR later.

ahatanak commented on clang/test/CodeGen/ptrauth-function-init.c:
```diff
@@ -0,0 +1,31 @@
+// CXX: store ptr getelementptr inbounds (i32, ptr ptrauth (ptr @f, i32 0), i64 2), ptr @_ZL2fp, align 8
+
+__attribute__((used))
+void (*const fp)(void) = (void (*)(void))((int *)&f + 2); // Error in C mode.
```
In C++, the global constructor can initialize the variable.

ahatanak commented on clang/test/CodeGen/ptrauth-weak_import.c:
```diff
@@ -0,0 +1,10 @@
+// RUN: %clang_cc1 -triple arm64-apple-ios -fptrauth-calls -fptrauth-intrinsics -emit-llvm %s  -o - | FileCheck %s
```
We can probably consolidate some of the tests. I'll look into it later.

kovdan01 replied:
Thanks. I'm OK with merging this "as is" and enhancing/re-organizing tests as a separate PR.

kovdan01 reviewed (approved):
LGTM, but this needs a rebase - the base branch was updated, and now the PR contains unrelated changes.

ahmedbougacha committed:
Adjust test/ConstantInt following rebase.

mikerice1969 commented on clang/include/clang/Basic/PointerAuthOptions.h:
```diff
@@ -14,10 +14,144 @@
+    assert((getOtherDiscrimination() != Discrimination::Constant ||
+            ConstantDiscriminatorOrNone) &&
+           "constant discrimination requires a constant!");
+    if (ConstantDiscriminatorOrNone)
+      ConstantDiscriminator = *ConstantDiscriminatorOrNone;
```
Hi @ahmedbougacha @ahatanak The static verifier is reporting that 'ConstantDiscriminator' is not initialized when taking the false branch of that if. Is that possible?

asl replied:
@mikerice1969 The logic is a bit convoluted here. But essentially this field is only initialized when `getOtherDiscrimination() == Discrimination::Constant`. And there is assert above for that and below in `getConstantDiscrimination()`. But probably it would be better to initialize it regardless.

hanickadot commented:
When this change is building at Compiler Explorer I noticed it spams build output with warnings:
```
In file included from /root/llvm-project/clang/include/clang/AST/Type.h:31,
                 from /root/llvm-project/clang/include/clang/AST/DeclarationName.h:16,
                 from /root/llvm-project/clang/include/clang/AST/DeclBase.h:19,
                 from /root/llvm-project/clang/include/clang/AST/Decl.h:20,
                 from /root/llvm-project/clang/tools/libclang/CXExtractAPI.cpp:20:
/root/llvm-project/clang/include/clang/Basic/PointerAuthOptions.h:61:8: warning: 'clang::PointerAuthSchema::TheKind' is too small to hold all values of 'enum class clang::PointerAuthSchema::Kind'
   61 |   Kind TheKind : 2;
      |        ^~~~~~~
/root/llvm-project/clang/include/clang/Basic/PointerAuthOptions.h:65:29: warning: 'clang::PointerAuthSchema::SelectedAuthenticationMode' is too small to hold all values of 'enum class clang::PointerAuthenticationMode'
   65 |   PointerAuthenticationMode SelectedAuthenticationMode : 2;
      |                             ^~~~~~~~~~~~~~~~~~~~~~~~~~
/root/llvm-project/clang/include/clang/Basic/PointerAuthOptions.h:66:18: warning: 'clang::PointerAuthSchema::DiscriminationKind' is too small to hold all values of 'enum class clang::PointerAuthSchema::Discrimination'
   66 |   Discrimination DiscriminationKind : 2;
      |                  ^~~~~~~~~~~~~~~~~~
      ```
compiler-explorar pipe line builds it with GCC 9.2

asl commented:
@ahatanak Will you please take a look into all these issues?

ahatanak commented:
Do we have to fix this just to appease gcc 9.2? It looks like newer versions of gcc (after 9.3) don't emit the warning.
